### PR TITLE
[Proof of Concept] Default tags implementation: provider and subnet/vpc resources

### LINKED ISSUE #17826: [Proof of Concept] provider: Default tags implementation
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Placeholder for provider-level implementation of `default_tags`. 

This includes:
  *  the updated provider schema including the "default_tags" configuration block
```hcl
provider "aws" {
default_tags {
    tags = {
      key1 = providervalue1
      key2 = providervalue2
    }  
    # ... potentially additional configurations in the future 
  }
}
```
 * the updated `Config` struct (where the Terraform AWS Provider saves provider configuration values before initializing AWS service clients) 
 * the updated `keyvaluetags` package to include the `DefaultConfig` struct and receiver methods
 * updated provider documentation (might be more applicable when addressing #17827 as the POC will be made available in 2 resources)

### New or Affected Resource(s)

* N/A

### References

* #7926

### LINKED ISSUE #17827: [Proof of Concept] resources: Default tags implementation
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Placeholder for initial resource-level implementation of default_tags.

This will include support for provider-wide `default_tags` within 2 key resources. Support includes:
* `CustomizeDiff` w/in resource definition to account for plan-time differences
* `tags_all` Computed attribute
* handling within CRUD operations
* acceptance test coverage for the 4 use cases presented in #7926 (provider-level only, resource-only, overlapping, and non-overlapping tags)

### New or Affected Resource(s)

* aws_subnet
* aws_vpc

### References

* #7926

### PR DESCRIPTION
**Pull request by anGie44**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #17826 
Closes #17827 

Output from acceptance testing:

```
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (15.28s)
--- PASS: TestAccAWSProvider_Region_AwsSC2S (15.50s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (15.53s)
--- PASS: TestAccAWSProvider_Region_AwsC2S (15.78s)
--- PASS: TestAccAWSProvider_Region_AwsChina (16.42s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_None (19.23s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_One (19.51s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (19.58s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (19.80s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (19.92s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (19.95s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (19.96s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_Multiple (19.97s)
--- PASS: TestAccAWSProvider_Endpoints (19.97s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (19.96s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (20.19s)
--- PASS: TestAccAWSProvider_DefaultAndIgnoreTags_EmptyConfigurationBlocks (20.24s)
--- PASS: TestAccAWSProvider_AssumeRole_Empty (22.44s)

--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_duplicateTag (1.90s)
--- PASS: TestAccAWSSubnet_defaultTags_updateToProviderOnly (36.35s)
--- PASS: TestAccAWSSubnet_defaultTags_updateToResourceOnly (39.64s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_nonOverlappingTag (48.80s)
--- PASS: TestAccAWSSubnet_defaultTags_providerOnly (50.92s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_overlappingTag (51.19s)
--- PASS: TestAccAWSSubnet_defaultAndIgnoreTags (45.00s)

--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_duplicateTag (2.27s)
--- PASS: TestAccAWSVpc_defaultTags_updateToProviderOnly (31.94s)
--- PASS: TestAccAWSVpc_defaultTags_updateToResourceOnly (33.68s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_overlappingTag (47.60s)
--- PASS: TestAccAWSVpc_defaultTags_providerOnly (47.63s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_nonOverlappingTag (47.75s)
--- PASS: TestAccAWSVpc_defaultAndIgnoreTags (29.82s)
```

### CODE REVIEW

**Comment by bflad** [comment]
Pending some potential changes noted over in https://github.com/hashicorp/terraform-provider-aws/pull/17941#pullrequestreview-612196815

File: aws/tags.go [hunk 1]
```diff
@@ -75,3 +78,42 @@ func ec2TagSpecificationsFromMap(m map[string]interface{}, t string) []*ec2.TagS
...
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	resourceTags := keyvaluetags.New(diff.Get("tags").(map[string]interface{}))
+
+	if defaultTagsConfig.TagsEqual(resourceTags) {
+		return fmt.Errorf(`"tags" are identical to those in the "default_tags" configuration block of the provider: please de-duplicate and try again`)
+	}
+
+	allTags := defaultTagsConfig.MergeTags(resourceTags).IgnoreConfig(ignoreTagsConfig)
```
Comment (anGie44) [inline aws/tags.go:112]: to accommodate cases where both (default and ignore are configured). do you think this is a valid use case?

Comment (bflad) [reply aws/tags.go:112]: I think configuring both is certainly valid as long as the tag keys do not overlap. For example, if I'm working in engineering and running Kubernetes I may want to do something like:

```terraform
provider "aws" {
  default_tags {
    tags = {
      CostCenter = "Engineering"
    }
  }

  ignore_tags {
    key_prefixes = ["kubernetes.io/"]
  }
}
```

Maybe we should error in the overlapping keys case, if we do not already. Does not need to be in the initial pull request though. 😄

**Review by bflad — COMMENTED** [review]
This looks really good, I'm stepping into a meeting for an hour, but all I have left to do is run the testing at this point to verify. 👍

**Review by bflad — APPROVED** [review]
Two nits on the changelog entries, otherwise this is a fantastic start. Great work. 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSProvider_AssumeRole_Empty (47.43s)
--- PASS: TestAccAWSProvider_DefaultAndIgnoreTags_EmptyConfigurationBlocks (23.95s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_Multiple (39.35s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_None (43.78s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_One (32.28s)
--- PASS: TestAccAWSProvider_Endpoints (40.52s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (30.92s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (38.83s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (31.18s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (32.62s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (39.85s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (35.49s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (36.41s)
--- PASS: TestAccAWSProvider_Region_AwsC2S (33.25s)
--- PASS: TestAccAWSProvider_Region_AwsChina (27.06s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (31.68s)
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (36.04s)
--- PASS: TestAccAWSProvider_Region_AwsSC2S (33.29s)

--- PASS: TestAccAWSSubnet_availabilityZoneId (62.10s)
--- PASS: TestAccAWSSubnet_basic (66.30s)
--- PASS: TestAccAWSSubnet_defaultAndIgnoreTags (172.35s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_duplicateTag (11.92s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_nonOverlappingTag (198.12s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_overlappingTag (197.36s)
--- PASS: TestAccAWSSubnet_defaultTags_providerOnly (185.18s)
--- PASS: TestAccAWSSubnet_defaultTags_updateToProviderOnly (135.01s)
--- PASS: TestAccAWSSubnet_defaultTags_updateToResourceOnly (136.87s)
--- PASS: TestAccAWSSubnet_disappears (71.35s)
--- PASS: TestAccAWSSubnet_enableIpv6 (123.25s)
--- PASS: TestAccAWSSubnet_ignoreTags (173.12s)
--- PASS: TestAccAWSSubnet_ipv6 (119.99s)
--- PASS: TestAccAWSSubnet_MapPublicIpOnLaunch (234.20s)
--- PASS: TestAccAWSSubnet_tags (140.78s)
--- SKIP: TestAccAWSSubnet_CustomerOwnedIpv4Pool (1.04s)
--- SKIP: TestAccAWSSubnet_MapCustomerOwnedIpOnLaunch (0.71s)
--- SKIP: TestAccAWSSubnet_outpost (1.02s)

--- PASS: TestAccAWSVpc_AssignGeneratedIpv6CidrBlock (161.38s)
--- PASS: TestAccAWSVpc_basic (84.09s)
--- PASS: TestAccAWSVpc_bothDnsOptionsSet (103.70s)
--- PASS: TestAccAWSVpc_classiclinkDnsSupportOptionSet (90.72s)
--- PASS: TestAccAWSVpc_classiclinkOptionSet (89.79s)
--- PASS: TestAccAWSVpc_coreMismatchedDiffs (73.00s)
--- PASS: TestAccAWSVpc_defaultAndIgnoreTags (84.82s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_duplicateTag (11.01s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_nonOverlappingTag (187.70s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_overlappingTag (88.86s)
--- PASS: TestAccAWSVpc_defaultTags_providerOnly (155.12s)
--- PASS: TestAccAWSVpc_defaultTags_updateToProviderOnly (131.24s)
--- PASS: TestAccAWSVpc_defaultTags_updateToResourceOnly (131.38s)
--- PASS: TestAccAWSVpc_DisabledDnsSupport (101.98s)
--- PASS: TestAccAWSVpc_disappears (57.68s)
--- PASS: TestAccAWSVpc_ignoreTags (93.11s)
--- PASS: TestAccAWSVpc_tags (211.03s)
--- PASS: TestAccAWSVpc_Tenancy (161.25s)
--- PASS: TestAccAWSVpc_update (150.61s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSProvider_AssumeRole_Empty (54.05s)
--- PASS: TestAccAWSProvider_DefaultAndIgnoreTags_EmptyConfigurationBlocks (43.18s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_Multiple (32.95s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_None (21.45s)
--- PASS: TestAccAWSProvider_DefaultTags_Tags_One (37.71s)
--- PASS: TestAccAWSProvider_Endpoints (40.39s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (38.03s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (31.33s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (32.15s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (35.19s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (40.27s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (32.84s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (42.61s)
--- PASS: TestAccAWSProvider_Region_AwsC2S (31.91s)
--- PASS: TestAccAWSProvider_Region_AwsChina (27.52s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (18.89s)
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (20.20s)
--- PASS: TestAccAWSProvider_Region_AwsSC2S (19.41s)

--- PASS: TestAccAWSSubnet_availabilityZoneId (53.68s)
--- PASS: TestAccAWSSubnet_basic (103.10s)
--- PASS: TestAccAWSSubnet_defaultAndIgnoreTags (174.14s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_duplicateTag (14.71s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_nonOverlappingTag (172.37s)
--- PASS: TestAccAWSSubnet_defaultTags_providerAndResource_overlappingTag (180.57s)
--- PASS: TestAccAWSSubnet_defaultTags_providerOnly (181.37s)
--- PASS: TestAccAWSSubnet_defaultTags_updateToProviderOnly (135.37s)
--- PASS: TestAccAWSSubnet_defaultTags_updateToResourceOnly (140.56s)
--- PASS: TestAccAWSSubnet_disappears (77.51s)
--- PASS: TestAccAWSSubnet_enableIpv6 (128.65s)
--- PASS: TestAccAWSSubnet_ignoreTags (176.63s)
--- PASS: TestAccAWSSubnet_ipv6 (121.96s)
--- PASS: TestAccAWSSubnet_MapPublicIpOnLaunch (179.33s)
--- PASS: TestAccAWSSubnet_tags (245.77s)
--- SKIP: TestAccAWSSubnet_CustomerOwnedIpv4Pool (1.99s)
--- SKIP: TestAccAWSSubnet_MapCustomerOwnedIpOnLaunch (0.81s)
--- SKIP: TestAccAWSSubnet_outpost (2.42s)

--- FAIL: TestAccAWSVpc_classiclinkDnsSupportOptionSet (34.06s) # https://github.com/hashicorp/terraform-provider-aws/issues/17460
--- FAIL: TestAccAWSVpc_classiclinkOptionSet (35.36s) # https://github.com/hashicorp/terraform-provider-aws/issues/17460
--- PASS: TestAccAWSVpc_AssignGeneratedIpv6CidrBlock (212.69s)
--- PASS: TestAccAWSVpc_basic (83.60s)
--- PASS: TestAccAWSVpc_bothDnsOptionsSet (99.42s)
--- PASS: TestAccAWSVpc_coreMismatchedDiffs (75.31s)
--- PASS: TestAccAWSVpc_defaultAndIgnoreTags (163.31s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_duplicateTag (6.40s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_nonOverlappingTag (175.66s)
--- PASS: TestAccAWSVpc_defaultTags_providerAndResource_overlappingTag (172.40s)
--- PASS: TestAccAWSVpc_defaultTags_providerOnly (146.26s)
--- PASS: TestAccAWSVpc_defaultTags_updateToProviderOnly (108.09s)
--- PASS: TestAccAWSVpc_defaultTags_updateToResourceOnly (115.28s)
--- PASS: TestAccAWSVpc_DisabledDnsSupport (100.36s)
--- PASS: TestAccAWSVpc_disappears (57.28s)
--- PASS: TestAccAWSVpc_ignoreTags (154.45s)
--- PASS: TestAccAWSVpc_tags (206.08s)
--- PASS: TestAccAWSVpc_Tenancy (214.24s)
--- PASS: TestAccAWSVpc_update (151.29s)
```

File: aws/resource_aws_subnet.go [hunk 2]
```diff
@@ -267,10 +274,17 @@ func resourceAwsSubnetRead(d *schema.ResourceData, meta interface{}) error {
 
 	d.Set("arn", subnet.SubnetArn)
 
-	if err := d.Set("tags", keyvaluetags.Ec2KeyValueTags(subnet.Tags).IgnoreAws().IgnoreConfig(ignoreTagsConfig).Map()); err != nil {
+	tags := keyvaluetags.Ec2KeyValueTags(subnet.Tags).IgnoreAws().IgnoreConfig(ignoreTagsConfig)
+
+	//lintignore:AWSR002
```
[the remark this answers is not in the document]
Comment (anGie44) [reply aws/resource_aws_subnet.go:279]: sounds good! i'll make an in issue for it and follow-up 👍

**Pull request merged**